docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        Callable = 1 << 2, // Indicates if an instance of the type can be called like a function. (It has a `__call__` method).
+        BuiltIn = 1 << 3, // Indicates if the type is a built-in type (like `int`, `str`, etc.).
+        Literal = 1 << 4, // Indicates if the instance is a literal (like `42`, `"hello"`, etc.).
+        Interface = 1 << 5, // Indicates if the type is an interface (a type that defines a set of methods and properties). In Python this would be a Protocol.
+        Generic = 1 << 6, // Indicates if the type is a generic type (a type that can be parameterized with other types).
+    }
+
+    // Flags that describe the characteristics of a function or method.
+    // These flags can be combined using bitwise operations.
+    export const enum FunctionFlags {
+        None = 0,
+        Async = 1 << 0, // Indicates if the function is asynchronous.
+        Generator = 1 << 1, // Indicates if the function is a generator (can yield values).
+        Abstract = 1 << 2, // Indicates if the function is abstract (must be implemented in a subclass).
+        Static = 1 << 3, // Indicates if the function is static (belongs to the class rather than an instance).
+    }
+
+    // Flags that describe the characteristics of a class.
+    // These flags can be combined using bitwise operations.
+    export const enum ClassFlags {
+        None = 0,
+        Enum = 1 << 0, // Indicates if the class is an enum (a special kind of class that defines a set of named values).
+    }
+
+    export interface Type {
+        // Unique identifier for the type definition within the session.
+        handle: string | number;
+
+        // Essential classification of the symbol.
+        category: TypeCategory;
+
+        // Flags describing the type.
+        flags: TypeFlags;
+
+        // Name of the module the type comes from
+        moduleName: string | undefined;
+
+        // Name of the type.
+        name: string;
+
+        // Flags specific to the category
+        categoryFlags: number;
+    }
+
+    export interface Member {
+        // The name of the member.
+        name: string;
+
+        // The type of the member.
+        type: Type;
+
+        // The type the member came from
+        classType: Type | undefined;
+    }
+
+    // Flags that are used for searching for members of a type.
+    export const enum MemberAccessFlags {
+        None = 0,
+        SkipInstanceMembers = 1 << 0, // Skip instance members when searching for members of a type.
+        SkipObjectBaseClass = 1 << 1, // Skip members from the `object` base class when searching for members of a type.
+    }
+
+    // Represents the category of a declaration in the type system.
+    // This is used to classify declarations such as variables, functions, classes, etc.
+    export const enum DeclarationCategory {
+        Intrinsic = 0, // An intrinsic is a built-in type or function that is part of the language itself, such as `int`, `str`, `len`, etc.
```

Comment:
This comment is a misleading explanation of what an "intrinsic" declaration is. After all, the symbols `int`, `str`, and `len` all have explicit type declarations in `builtins.pyi`. An intrinsic declaration is one that has no explicit declaration within the code. An example is the `__class__` symbol within a class body. Pyright has also historically used intrinsic declarations for certain dundered symbols that are omitted from the typeshed stubs, but those stubs have become more complete over time, so the need for intrinsic declarations has decreased.

Reply:
I'll fix the comment. Maybe we don't need this anymore? It is used by Pylance code but I think these could just be represented by the other categories?